### ISSUE TITLE
False positive NS1004 with `When().Do()`

### ISSUE THREAD

**Issue by mcarton** [nsubstitute/NSubstitute.Analyzers#160]
The following gives a NS1004 warning, even though the test works as expected:

```cs
using NUnit.Framework;
using NSubstitute;
using System;

namespace dlsfkn
{
    public interface IFoo
    {
        string Foo { set; }
    }

    public class Tests
    {
        [Test]
        public void Test1()
        {
            var foo = NSubstitute.Substitute.For<IFoo>();

            foo
                .When(x => x.Foo = Arg.Any<string>())
                .Do(x => throw new Exception("This does fail"));

            foo.Foo = "whynot";
        }
    }
}
```

gives the following warning:

```
/tmp/dlsfkn/UnitTest1.cs(20,36): warning NS1004: Argument matcher used with a non-virtual member of a class. [/tmp/dlsfkn/dlsfkn.csproj]
```

with the following project:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>

    <IsPackable>false</IsPackable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="NUnit" Version="3.12.0" />
    <PackageReference Include="NUnit3TestAdapter" Version="3.16.1" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.5.0"/>

    <PackageReference Include="NSubstitute" Version="4.2.2" />
    <PackageReference Include="NSubstitute.Analyzers.CSharp" Version="1.0.14" />
  </ItemGroup>

</Project>
```

---

This is probably related to  #159, but with a different method, so I though I'd report it anyway.

**Comment by tpodolak** [comment]
Hi @mcarton thanks for reporting. As you said this is related with #159. @dtchepak could you confirm that this is correct usage of `Arg.Any`? I didn't find any usage examples of Arg.Any Arg.Do etc in docs.

**Comment by mcarton** [comment]
I also tried to test that a property was not set using the following, which doesn't use `Do`. This doesn't work (the test passes), but I would have expected it to work. It also gives a NS1004, even though `Foo` is from an interface:

```cs
using NUnit.Framework;
using NSubstitute;

namespace dlsfkn
{
    public interface IFoo
    {
        string Foo { set; }
    }

    public class Tests
    {
        [Test]
        public void Test1()
        {
            var foo = NSubstitute.Substitute.For<IFoo>();

            foo.DidNotReceive().Foo = Arg.Any<string>();

            foo.Foo = "whynot";
        }
    }
}
```

---

> @dtchepak could you confirm that this is correct usage of Arg.Any? I didn't find any usage examples of Arg.Any Arg.Do etc in docs.

What would you recommend instead to test that a property is not set?

**Comment by dtchepak** [comment]
@tpodolak I haven't expected that use but it seems like it is valid.

> What would you recommend instead to test that a property is not set?

@mcarton If possible I think asserting on the property value is cleaner. Alternatively using an explicit setter method if the act of setting is significant. But these are more subjective design opinions rather than anything concrete. 😄

**Comment by tpodolak** [comment]
Fixed in #169
